### PR TITLE
generate go wrapper for registry 1.3, handles pause/unpause upkeeps and check data updated events

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

File: contracts/scripts/native_solc_compile_all [hunk 1]
```diff
@@ -39,6 +39,8 @@ $SCRIPTPATH/native_solc7_compile tests/UpkeepCounter.sol
 $SCRIPTPATH/native_solc8_6_compile factories/CronUpkeepFactory.sol
 $SCRIPTPATH/native_solc8_6_compile KeeperRegistry.sol
 $SCRIPTPATH/native_solc8_6_compile dev/KeeperRegistryCheckUpkeepGasUsageWrapper.sol
+$SCRIPTPATH/native_solc8_6_compile dev/KeeperRegistryDev.sol
```
Comment (infiloop2) [inline contracts/scripts/native_solc_compile_all:42]: Rename to KeeperRegistry1_3

File: contracts/scripts/native_solc_compile_all [hunk 2]
```diff
@@ -39,6 +39,8 @@ $SCRIPTPATH/native_solc7_compile tests/UpkeepCounter.sol
...
+$SCRIPTPATH/native_solc8_6_compile dev/KeeperRegistryLogic.sol
```
Comment (infiloop2) [inline contracts/scripts/native_solc_compile_all:43]: This shouldn't be needed

File: core/services/keeper/common.go [hunk 3]
```diff
@@ -4,19 +4,22 @@ import (
 	"math/big"
 	"time"
 
-	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common"
 
-	evmtypes "github.com/smartcontractkit/chainlink/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/core/gethwrappers/generated/keeper_registry_wrapper1_3"
```
Comment (infiloop2) [inline core/services/keeper/common.go:9]: nit: can you club this import with the other wrapper imports

File: core/services/keeper/common.go [hunk 4]
```diff
@@ -4,19 +4,22 @@ import (
...
+	"github.com/smartcontractkit/chainlink/core/chains/evm/types"
 	"github.com/smartcontractkit/chainlink/core/gethwrappers/generated/keeper_registry_wrapper1_1"
 	"github.com/smartcontractkit/chainlink/core/gethwrappers/generated/keeper_registry_wrapper1_2"
 )
 
-var Registry1_1ABI = evmtypes.MustGetABI(keeper_registry_wrapper1_1.KeeperRegistryABI)
-var Registry1_2ABI = evmtypes.MustGetABI(keeper_registry_wrapper1_2.KeeperRegistryABI)
+var Registry1_1ABI = types.MustGetABI(keeper_registry_wrapper1_1.KeeperRegistryABI)
+var Registry1_2ABI = types.MustGetABI(keeper_registry_wrapper1_2.KeeperRegistryABI)
+var Registry1_3ABI = types.MustGetABI(keeper_registry_wrapper1_3.KeeperRegistryDevABI)
```
Comment (infiloop2) [inline core/services/keeper/common.go:18]: nit. rename KeeperRegistryDevABI to KeeperRegistryABI

File: core/services/keeper/orm.go [hunk 5]
```diff
@@ -265,3 +265,14 @@ func (korm ORM) SetLastRunInfoForUpkeepOnJob(jobID int32, upkeepID *utils.Big, h
...
+
+// UpdateCheckData updates check data for a job
+func (korm ORM) UpdateCheckData(upkeepId *utils.Big, jobId int32, checkData []byte) error {
+	stmt := `
+UPDATE upkeep_registrations
+SET check_data = $1
+WHERE upkeep_id = $2 AND registry_id = (SELECT id FROM keeper_registries WHERE job_id = $3)
+`
+	_, err := korm.q.Exec(stmt, checkData, upkeepId, jobId)
+	return errors.Wrapf(err, "failed to update check data for job id %s", upkeepId.String())
```
Comment (infiloop2) [inline core/services/keeper/orm.go:277]: nit. "check data for upkeep id %s"

File: core/services/keeper/registry_interface.go [hunk 6]
```diff
@@ -156,19 +197,19 @@ func (rw *RegistryWrapper) GetActiveUpkeepIDs(opts *bind.CallOpts) ([]*big.Int,
...
+			maxCount := upkeepCount.Int64() - int64(len(activeUpkeepIDs))
 			if maxCount > ActiveUpkeepIDBatchSize {
 				maxCount = ActiveUpkeepIDBatchSize
 			}
-			activeUpkeepIDBatch, err := rw.contract1_2.GetActiveUpkeepIDs(opts, big.NewInt(startIndex), big.NewInt(maxCount))
+			if rw.Version == RegistryVersion_1_2 {
+				activeUpkeepIDBatch, err = rw.contract1_2.GetActiveUpkeepIDs(opts, big.NewInt(startIndex), big.NewInt(maxCount))
+			} else {
+				activeUpkeepIDBatch, err = rw.contract1_3.GetActiveUpkeepIDs(opts, big.NewInt(startIndex), big.NewInt(maxCount))
+			}
```
Comment (infiloop2) [inline core/services/keeper/registry_interface.go:212]: nit. can you refactor the GetActiveUpkeepIDs logic to structure this better.

One way i can think of is a separate function that takes a contract wrapper as argument, with the wrapper satisfying interface with GetState, GetActiveUpkeepIDs functions, and executes the logic within it.

It will also eliminate the need for a separate getUpkeepCount function i think

cc @EasterTheBunny may have more thoughts on how to structure this

File: core/services/keeper/registry_interface_logs.go [hunk 7]
```diff
@@ -110,49 +141,112 @@ func (rw *RegistryWrapper) ParseUpkeepPerformedLog(broadcast log.Broadcast) (*Up
...
+			return nil, errors.Errorf("expected UpkeepMigrated log but got %T", broadcastedLog)
+		}
+		return broadcastedLog.Id, nil
 	default:
 		return nil, newUnsupportedVersionError("GetUpkeepIdFromMigratedLog", rw.Version)
 	}
 }
+
+func (rw *RegistryWrapper) GetUpkeepIdFromUpkeepPausedLog(broadcast log.Broadcast) (*big.Int, error) {
+	// Only supported on 1.3
```
Comment (infiloop2) [inline core/services/keeper/registry_interface_logs.go:219]: 👍

File: core/services/keeper/registry_synchronizer_log_listener.go [hunk 8]
```diff
@@ -33,31 +34,46 @@ func (rs *RegistrySynchronizer) HandleLog(broadcast log.Broadcast) {
...
+		*registry1_3.KeeperRegistryDevUpkeepReceived:
 		wasOverCapacity = rs.mailRoom.mbUpkeepReceived.Deliver(broadcast)
 		mailboxName = "mbUpkeepReceived"
-	case *registry1_2.KeeperRegistryUpkeepMigrated:
+	case *registry1_2.KeeperRegistryUpkeepMigrated,
+		*registry1_3.KeeperRegistryDevUpkeepMigrated:
 		wasOverCapacity = rs.mailRoom.mbUpkeepMigrated.Deliver(broadcast)
 		mailboxName = "mbUpkeepMigrated"
+	case *registry1_3.KeeperRegistryDevUpkeepPaused:
+		svcLogger.Debug("delivered to upkeep paused")
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_log_listener.go:71]: nit. I like having more debug logs, could you add them to all the different types :D

File: core/services/keeper/registry_synchronizer_core.go [hunk 9]
```diff
@@ -62,13 +65,15 @@ type RegistrySynchronizer struct {
...
-		mbUpkeepMigrated:    utils.NewMailbox[log.Broadcast](500),
-		mbUpkeepGasLimitSet: utils.NewMailbox[log.Broadcast](500),
+		mbUpkeepCanceled:         utils.NewMailbox[log.Broadcast](500),
+		mbSyncRegistry:           utils.NewMailbox[log.Broadcast](1),
+		mbUpkeepPerformed:        utils.NewMailbox[log.Broadcast](3000),
+		mbUpkeepRegistered:       utils.NewMailbox[log.Broadcast](500),
+		mbUpkeepReceived:         utils.NewMailbox[log.Broadcast](500),
+		mbUpkeepMigrated:         utils.NewMailbox[log.Broadcast](500),
+		mbUpkeepGasLimitSet:      utils.NewMailbox[log.Broadcast](500),
+		mbUpkeepPaused:           utils.NewMailbox[log.Broadcast](500),
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_core.go:75]: we'll also need to listen to Unpause event

File: go.mod [hunk 10]
```diff
@@ -81,15 +81,22 @@ require (
 	go.uber.org/zap v1.21.0
 	golang.org/x/crypto v0.0.0-20220722155217-630584e8d5aa
 	golang.org/x/exp v0.0.0-20220608143224-64259d1afd70
-	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
+	golang.org/x/sync v0.0.0-20220722155255-886fb9371eb4
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
 	golang.org/x/text v0.3.7
-	golang.org/x/tools v0.1.10
+	golang.org/x/tools v0.1.12
```
Comment (infiloop2) [inline go.mod:87]: are these changes related to your PR?

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 11]
```diff
@@ -309,3 +311,86 @@ func (rs *RegistrySynchronizer) handleUpkeepMigrated(broadcast log.Broadcast) {
...
+		rs.logger.Error(errors.Wrap(err, "Unable to fetch upkeep ID from upkeep paused log"))
+		return
+	}
+
+	affected, err := rs.orm.BatchDeleteUpkeepsForJob(rs.job.ID, []utils.Big{*utils.NewBig(pausedUpkeepId)})
+	if err != nil {
+		rs.logger.Error(errors.Wrap(err, "unable to batch delete upkeeps"))
+		return
+	}
+	rs.logger.Debugw(fmt.Sprintf("paused %v upkeep registrations", affected), "txHash", txHash)
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_process_logs.go:349]: nit. also log the upkeep ID here?

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 12]
```diff
@@ -309,3 +311,86 @@ func (rs *RegistrySynchronizer) handleUpkeepMigrated(broadcast log.Broadcast) {
...
+		return
+	}
+
+	err = rs.orm.UpdateCheckData(utils.NewBig(updateLog.UpkeepID), rs.job.ID, updateLog.NewCheckData)
+	if err != nil {
+		rs.logger.Error(errors.Wrapf(err, "unable to update check data for upkeep %s", updateLog.UpkeepID.String()))
+		return
+	}
+
+	rs.logger.Debugw(fmt.Sprintf("updated check data with %T", updateLog), "txHash", txHash)
```
Comment (infiloop2) [inline core/services/keeper/registry_synchronizer_process_logs.go:391]: Can we not log the full updateLog here? only upkeep ID would be better, can discuss reasoning on slack

File: core/services/keeper/registry1_3_synchronizer_test.go [hunk 13]
```diff
@@ -0,0 +1,691 @@
+package keeper_test
```
Comment (infiloop2) [inline core/services/keeper/registry1_3_synchronizer_test.go:1]: Love these tests 🎉

**Review by infiloop2 — CHANGES_REQUESTED** [review]
Overall looking great! Added some inline comments

File: contracts/scripts/native_solc_compile_all (hunk 1 above)
Comment (FelixFan1992) [reply contracts/scripts/native_solc_compile_all:42]: Done

File: contracts/scripts/native_solc_compile_all (hunk 2 above)
Comment (FelixFan1992) [reply contracts/scripts/native_solc_compile_all:43]: Done

File: core/services/keeper/common.go (hunk 4 above)
Comment (FelixFan1992) [reply core/services/keeper/common.go:18]: Done

File: core/services/keeper/common.go (hunk 3 above)
Comment (FelixFan1992) [reply core/services/keeper/common.go:9]: Done

File: core/services/keeper/registry_synchronizer_core.go (hunk 9 above)
Comment (FelixFan1992) [reply core/services/keeper/registry_synchronizer_core.go:75]: Done

File: core/services/keeper/orm.go (hunk 5 above)
Comment (FelixFan1992) [reply core/services/keeper/orm.go:277]: Done

File: core/services/keeper/orm.go [hunk 14]
```diff
@@ -265,3 +265,14 @@ func (korm ORM) SetLastRunInfoForUpkeepOnJob(jobID int32, upkeepID *utils.Big, h
 	last_run_block_height <= $1`, height, upkeepID, jobID, fromAddress.Hex())
 	return errors.Wrap(err, "SetLastRunInfoForUpkeepOnJob failed")
 }
+
+// UpdateCheckData updates check data for a job
+func (korm ORM) UpdateCheckData(upkeepId *utils.Big, jobId int32, checkData []byte) error {
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply core/services/keeper/orm.go:270]: Done

File: core/services/keeper/registry_synchronizer_log_listener.go (hunk 8 above)
Comment (FelixFan1992) [reply core/services/keeper/registry_synchronizer_log_listener.go:71]: Done

File: core/services/keeper/registry_synchronizer_process_logs.go (hunk 11 above)
Comment (FelixFan1992) [reply core/services/keeper/registry_synchronizer_process_logs.go:349]: Done. removed affected since it will also be 1.

File: core/services/keeper/registry_synchronizer_process_logs.go (hunk 12 above)
Comment (FelixFan1992) [reply core/services/keeper/registry_synchronizer_process_logs.go:391]: Done

File: go.mod (hunk 10 above)
Comment (FelixFan1992) [reply go.mod]: remove

File: core/services/keeper/registry_interface.go (hunk 6 above)
Comment (FelixFan1992) [reply core/services/keeper/registry_interface.go:212]: the problem with this is
in 1.1, we use `GetUpkeepCount` to get upkeep count,
in 1.2, we use `GetState` to get a `keeper_registry_wrapper1_**2**.GetState` object and then access the upkeep count.
in 1.3, we use `GetState` to get a `keeper_registry_wrapper1_**3**.GetState` object and then access the upkeep count. 
this makes it difficult to use a uniform interface.

similar to that, in `GetActiveUpkeepIDs`, in 1.1 we need to call `GetCanceledUpkeepList` so it cannot be in the same interface.

i feel that makes the refacotring hard. let me know what you think

File: core/services/keeper/registry1_3_synchronizer_test.go [hunk 15]
```diff
@@ -0,0 +1,720 @@
...
+	require.Contains(t, logListenerOpts.LogsWithTopics, registry1_3.KeeperRegistryUpkeepCanceled{}.Topic(), "Registry should listen to KeeperRegistryUpkeepCanceled log")
+	require.Contains(t, logListenerOpts.LogsWithTopics, registry1_3.KeeperRegistryUpkeepRegistered{}.Topic(), "Registry should listen to KeeperRegistryUpkeepRegistered log")
+	require.Contains(t, logListenerOpts.LogsWithTopics, registry1_3.KeeperRegistryUpkeepPerformed{}.Topic(), "Registry should listen to KeeperRegistryUpkeepPerformed log")
+	require.Contains(t, logListenerOpts.LogsWithTopics, registry1_3.KeeperRegistryUpkeepGasLimitSet{}.Topic(), "Registry should listen to KeeperRegistryUpkeepGasLimitSet log")
+	require.Contains(t, logListenerOpts.LogsWithTopics, registry1_3.KeeperRegistryUpkeepMigrated{}.Topic(), "Registry should listen to KeeperRegistryUpkeepMigrated log")
+	require.Contains(t, logListenerOpts.LogsWithTopics, registry1_3.KeeperRegistryUpkeepReceived{}.Topic(), "Registry should listen to KeeperRegistryUpkeepReceived log")
+	require.Contains(t, logListenerOpts.LogsWithTopics, registry1_3.KeeperRegistryUpkeepPaused{}.Topic(), "Registry should listen to KeeperRegistryUpkeepPaused log")
```
Comment (infiloop2) [inline core/services/keeper/registry1_3_synchronizer_test.go:118]: also unpause

File: core/services/keeper/registry_interface.go [hunk 16]
```diff
@@ -140,10 +180,11 @@ func (rw *RegistryWrapper) GetActiveUpkeepIDs(opts *bind.CallOpts) ([]*big.Int,
 			}
 		}
 		return activeUpkeeps, nil
-	case RegistryVersion_1_2:
+	case RegistryVersion_1_2, RegistryVersion_1_3:
 		if opts == nil || opts.BlockNumber.Int64() == 0 {
+			var header *types.Header
 			// fetch the current block number so batched GetActiveUpkeepIDs calls can be performed on the same block
-			header, err := rw.evmClient.HeaderByNumber(context.Background(), nil)
+			header, err = rw.evmClient.HeaderByNumber(context.Background(), nil)
```
Comment (infiloop2) [inline core/services/keeper/registry_interface.go:187]: We also need to pass this opt in the below getState call (which is now refactored into getUpkeepCount call)

Can you move this code block to the top of the function? Let's use the consistent block number in all eth calls

File: core/services/keeper/registry_interface.go (hunk 6 above)
Comment (infiloop2) [reply core/services/keeper/registry_interface.go:212]: I was more thinking of a common interface for 1.2 and 1.3 since they have the same logic and same function calls, but it might be an overkill. Not a blocker imo, will leave it upto you

File: core/services/keeper/registry_interface_logs.go [hunk 17]
```diff
@@ -42,6 +43,24 @@ func (rw *RegistryWrapper) GetLogListenerOpts(minIncomingConfirmations uint32, u
...
+			LogsWithTopics: map[common.Hash][][]log.Topic{
+				registry1_3.KeeperRegistryKeepersUpdated{}.Topic():         nil,
+				registry1_3.KeeperRegistryConfigSet{}.Topic():              nil,
+				registry1_3.KeeperRegistryUpkeepCanceled{}.Topic():         nil,
+				registry1_3.KeeperRegistryUpkeepRegistered{}.Topic():       nil,
+				registry1_3.KeeperRegistryUpkeepPerformed{}.Topic():        upkeepPerformedFilter,
+				registry1_3.KeeperRegistryUpkeepGasLimitSet{}.Topic():      nil,
+				registry1_3.KeeperRegistryUpkeepMigrated{}.Topic():         nil,
+				registry1_3.KeeperRegistryUpkeepReceived{}.Topic():         nil,
+				registry1_3.KeeperRegistryUpkeepPaused{}.Topic():           nil,
```
Comment (infiloop2) [inline core/services/keeper/registry_interface_logs.go:59]: unpause

**Review by infiloop2 — CHANGES_REQUESTED** [review]
Some small comments remaining. Also could you update the title and summary of the PR please

File: core/services/keeper/registry1_3_synchronizer_test.go (hunk 15 above)
Comment (FelixFan1992) [reply core/services/keeper/registry1_3_synchronizer_test.go:118]: Done

File: core/services/keeper/registry_interface.go (hunk 16 above)
Comment (FelixFan1992) [reply core/services/keeper/registry_interface.go:187]: Good catch

File: core/services/keeper/registry_interface_logs.go (hunk 17 above)
Comment (FelixFan1992) [reply core/services/keeper/registry_interface_logs.go:59]: Done. I think log listener does not affect mailroom delivery and retrieval.

File: core/services/keeper/registry_synchronizer_process_logs.go [hunk 18]
```diff
@@ -309,3 +312,132 @@ func (rs *RegistrySynchronizer) handleUpkeepMigrated(broadcast log.Broadcast) {
...
+		return
+	}
+
+	updateLog, err := rs.registryWrapper.ParseUpkeepCheckDataUpdatedLog(broadcast)
+	if err != nil {
+		rs.logger.Error(errors.Wrap(err, "Unable to parse update log from upkeep check data updated log"))
+		return
+	}
+
+	err = rs.orm.UpdateCheckData(utils.NewBig(updateLog.UpkeepID), rs.job.ID, updateLog.NewCheckData)
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply core/services/keeper/registry_synchronizer_process_logs.go:432]: Done. I like it.

**Comment by FelixFan1992** [comment]
This PR implements 3 stories: [1](https://app.shortcut.com/chainlinklabs/story/49802/add-go-wrapper-and-enum-value-for-v1-3-registry), [2](https://app.shortcut.com/chainlinklabs/story/49812/support-pause-unpause-events), [3](https://app.shortcut.com/chainlinklabs/story/49813/support-checkdata-mutation-event)

It generates go wrapper for registry 1.3, handles pause/unpause upkeep, check data updated events.

File: core/services/keeper/registry1_1_synchronizer_test.go [hunk 19]
```diff
@@ -55,6 +55,8 @@ func mockRegistry1_1(
 ) {
 	registryMock := cltest.NewContractMockReceiver(t, ethMock, keeper.Registry1_1ABI, contractAddress)
 
+	ethMock.On("HeaderByNumber", mock.Anything, mock.Anything).
+		Return(&types.Header{Number: big.NewInt(10)}, nil)
```
Comment (infiloop2) [inline core/services/keeper/registry1_1_synchronizer_test.go:59]: curious, is this needed? do your changes affect registry1_1synchronizer?

Comment (FelixFan1992) [reply core/services/keeper/registry1_1_synchronizer_test.go:59]: this is bc we moved the check for opts *bind.CallOpts in `GetActiveUpkeepIDs` but in 1.1 test, we did not mock this call.

File: core/services/keeper/registry_interface_logs.go (hunk 17 above)
Comment (infiloop2) [reply core/services/keeper/registry_interface_logs.go:59]: hmm yeah, this would be covered by integration test which we should add as a followup

**Comment by infiloop2** [comment]
Looks good, thanks for addressing all the comments @FelixFan1992 ! Not accepting for now until we decide to merge it

**Review by infiloop2 — APPROVED** [review]
